models/src/main/scala/coop/rchain/models/rholangn/parmanager/ProtobufSerialization.scala

```diff
@@ -0,0 +1,66 @@
+package coop.rchain.models.rholangn.parmanager
+
+import cats.Eval
+import com.google.protobuf.{CodedInputStream, CodedOutputStream}
+import coop.rchain.models.rholangn.{ParN, RhoTypeN}
+
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream}
+import scala.util.Using
+
+object ProtobufSerialization {
+
+  @SuppressWarnings(Array("org.wartremover.warts.TryPartial"))
+  def serialize(par: RhoTypeN): Eval[Array[Byte]] = {
+
+    /** Wrapper for protobuf serialization of primitive types. */
+    class ProtobufPrimitiveWriter(output: CodedOutputStream) extends PrimitiveWriter {
+      def write(x: Byte): Eval[Unit] = Eval.later(output.writeRawByte(x))
+
+      /** Writes raw bytes without size prefix */
+      def writeRaw(x: Array[Byte]): Eval[Unit] = Eval.later(output.writeRawBytes(x))
+
+      /** Writes bytes with size prefix */
+      def write(x: Array[Byte]): Eval[Unit] = Eval.later(output.writeByteArrayNoTag(x))
+      def write(x: Boolean): Eval[Unit]     = Eval.later(output.writeBoolNoTag(x))
+      def write(x: Int): Eval[Unit]         = Eval.later(output.writeUInt32NoTag(x))
+      def write(x: Long): Eval[Unit]        = Eval.later(output.writeUInt64NoTag(x))
+      def write(x: String): Eval[Unit]      = Eval.later(output.writeStringNoTag(x))
+    }
+
+    // TODO: Properly handle errors
+    par.serializedSize.flatMap { serSize =>
+      Using(new ByteArrayOutputStream(serSize)) { baos =>
+        val cos    = CodedOutputStream.newInstance(baos)
+        val writer = new ProtobufPrimitiveWriter(cos)
+        Serialization.serialize(par, writer).map { _ =>
+          baos.flush()
+          val r = baos.toByteArray
+          cos.flush()
```

Comment:
After `baos.toByteArray` the read from output stream is done. So call to flush should be before getting final bytes, and to be 100% correct before outer ByteArrayOutputStream.